pkg/datapath/linux/node.go

```diff
@@ -855,7 +855,7 @@ func (n *linuxNodeHandler) enableIPsec(newNode *nodeTypes.Node) {
 
 				/* Insert wildcard policy rules for traffic skipping back through host */
 				ipsecIPv4Wildcard := &net.IPNet{IP: net.ParseIP(wildcardIPv4), Mask: net.IPv4Mask(0, 0, 0, 0)}
-				if err = ipsec.IpSecReplacePolicyFwd(ipsecIPv4Wildcard, ipsecRemote, ipsecLocal, ipsecRemote); err != nil {
+				if err = ipsec.IpSecReplacePolicyFwd(ipsecIPv4Wildcard, ipsecRemote, ipsecIPv4Wildcard, ipsecRemote); err != nil {
```

Comment:
So if I understand correctly, we have two pieces of code calling into `IpsecReplacePolicyFwd` with different values for the `tmplSrc`. That causes the `fwd` policy to change and packets to drop.
- Do we know if it's the back and forth change itself that causes the drop or one of the two `tmplSrc` values that's incorrect?
- Why did it change back and forth between the two values? As far as I can see, the two pieces of code should only be called once on startup :confused: 
- In `pkg/datapath/linux/node.go`, I can see a lot of calls to `UpsertIPsecEndpoint` with different values for the `tmplSrc` argument. If the issue is the difference in values for `tmplSrc` between (1) and (2), how can we be sure it won't happen again when some other `UpsertIPsecEndpoint` call in `pkg/datapath/linux/node.go` is made (for example when node encryption is enabled)? They don't all use `ipsecIPv4Wildcard` for `tmplSrc`.

Reply:
> > Do we know if it's the back and forth change itself that causes the drop or one of the two tmplSrc values that's incorrect?
> 
> We do not. Initially, I thought it was the changes but now that I've written everything up, I'm not so sure anymore.

It should be possible to confirm that by fixing the code to consistently use the other value (not wildcard) and see if it still fixes the bug.

I also suspect it's the change that's causing this, which is a bit worrying. If changing xfrm policies at runtime can cause packets drops then we might have other bugs.

> > Why did it change back and forth between the two values? As far as I can see, the two pieces of code should only be called once on startup
> 
> AFAIU, they are also called in the node update handler. Specifically, they are called in [enableIPsec](https://github.com/cilium/cilium/blob/2022e00724db3f889c22a151e5e5e806713e197a/pkg/datapath/linux/node.go#L856), which is called by [nodeUpdate](https://github.com/cilium/cilium/blob/2022e00724db3f889c22a151e5e5e806713e197a/pkg/datapath/linux/node.go#L939), which is called in [NodeAdd](https://github.com/cilium/cilium/blob/2022e00724db3f889c22a151e5e5e806713e197a/pkg/datapath/linux/node.go#L439), [NodeUpdate](https://github.com/cilium/cilium/blob/2022e00724db3f889c22a151e5e5e806713e197a/pkg/datapath/linux/node.go#L452) and [NodeConfigurationChanged](https://github.com/cilium/cilium/blob/2022e00724db3f889c22a151e5e5e806713e197a/pkg/datapath/linux/node.go#L1455). Am I missing something?

I had missed the latest. And it seems it's [called from Reinitialize](https://github.com/cilium/cilium/blob/2022e00724db3f889c22a151e5e5e806713e197a/pkg/datapath/loader/base.go#L441) every time we reload the base datapath, so that could explain the changes.

> > If the issue is the difference in values for tmplSrc between (1) and (2), how can we be sure it won't happen again when some other UpsertIPsecEndpoint call in pkg/datapath/linux/node.go is made (for example when node encryption is enabled)?
> 
> The forwarding policy on `UpsertIPsecEndpoint` is only added when [dir == IPSecDirIn || dir == IPSecDirBoth](https://github.com/cilium/cilium/blob/2022e00724db3f889c22a151e5e5e806713e197a/pkg/datapath/linux/ipsec/ipsec_linux.go#L343). At least for `enableIPsec` and [n.nodeConfig.EnableIPv4 && newNode.IPv4AllocCIDR != nil](https://github.com/cilium/cilium/blob/2022e00724db3f889c22a151e5e5e806713e197a/pkg/datapath/linux/node.go#L864-L890), we will always add a forwarding rule with the wildcard AFAICT. This seems to be the only case when we add an extra forwarding rule as introduced by [0b52fd7](https://github.com/cilium/cilium/commit/0b52fd76c0101e966d701c07ca174517948739e4).

Ok :+1: I was mostly worried about calls made in [`enableSubnetIPsec`](https://github.com/cilium/cilium/blob/a9abb1c8f2c206ad89f5bfcc28e57ab6befba113/pkg/datapath/linux/node.go#L541) but it seems we [can never call both](https://github.com/cilium/cilium/blob/a9abb1c8f2c206ad89f5bfcc28e57ab6befba113/pkg/datapath/linux/node.go#L1046) `enableSubnetIPsec` and `enableIPsec`.

Another reason for raising this is that I'm wondering if it really makes sense to have several places where we install `fwd` policies. Shouldn't we ensure we install this policy from a single function rather than having to ensure the `tmplSrc` is consistent between all functions?

Reply:
> Shouldn't we ensure we install this policy from a single function rather than having to ensure the `tmplSrc` is consistent between all functions?

I thought about this approach (as well as the approach of inserting two different templates), but I did not go for it because the two different forward policies are generated by different events. Looking at: https://github.com/cilium/cilium/blob/2022e00724db3f889c22a151e5e5e806713e197a/pkg/datapath/linux/node.go#L864-L890

One is added if `newNode.IsLocal()` and a rule is upserted with `IPSecDirIn`, while the other if `!newNode.isLocal()` and a rule is upserted with `IPSecDirOut`. So they way things work now where we insert policies by reacting to node updates, I think we need to insert them.

Another issue that we have now is that, even if the  policies do _not_ change, we still update them when we receive node updates. In some ways, that seems to be a good property to have because i) we do not keep any local state and ii) if something changes in the xfrm policies under our noses, things will work as best as they can. I do wonder, however, if there is a way better way to do things wrt when we  update the xfrm policies.

Reply:
> It should be possible to confirm that by fixing the code to consistently use the other value (not wildcard) and see if it still fixes the bug.

Based on my testing, it seems that the issue was the non-wildcard template rather than the switching of policies. I updated the commit message to reflect that.

Since the actual code has already went through CI twice, I would opt for marking this ready-to-merge (see https://github.com/cilium/cilium/pull/17865#issuecomment-969032164 for a discussion of CI failures) unless there are any  objections.